cpp/Platform.Delegates/Platform.Delegates.MulticastDelegate.h

```diff
@@ -126,57 +88,98 @@ namespace Platform::Delegates
 
         MulticastDelegate &operator+=(const DelegateType &callback)
         {
-            const std::lock_guard<std::mutex> lock(mutex);
+            std::lock_guard lock(mutex);
             this->callbacks.emplace_back(callback);
             return *this;
         }
 
-        MulticastDelegate &operator+=(DelegateRawFunctionType &callback)
+        MulticastDelegate &operator+=(DelegateRawFunctionType callback)
         {
-            return *this += DelegateType(callback);
+            return *this += DelegateType{callback};
         }
 
         MulticastDelegate &operator+=(const DelegateFunctionType &callback)
         {
-            return *this += DelegateType(callback);
+            return *this += DelegateType{callback};
         }
 
         MulticastDelegate &operator-=(const DelegateType &callback)
         {
-            const std::lock_guard<std::mutex> lock(mutex);
+            const std::lock_guard lock{mutex};
             auto searchResult = std::find(this->callbacks.rbegin(), this->callbacks.rend(), callback);
             if (searchResult != this->callbacks.rend()) {
-                auto removedCallbackPosition = --(searchResult.base());
-                this->callbacks.erase(removedCallbackPosition);
+                this->callbacks.erase(searchResult.base());
             }
             return *this;
         }
 
         MulticastDelegate &operator-=(DelegateRawFunctionType &callback)
         {
-            return *this -= DelegateType(callback);
+            return *this -= DelegateType{callback};
         }
 
         MulticastDelegate &operator-=(const DelegateFunctionType &callback)
         {
-            return *this -= DelegateType(callback);
+            return *this -= DelegateType{callback};
         }
 
-        virtual ReturnType operator()(Args... args) override
+        ReturnType operator()(Args... args) override
         {
-            const std::lock_guard<std::mutex> lock(mutex);
-            auto callbacksSize = this->callbacks.size();
-            if (callbacksSize == 0)
+            const std::lock_guard lock(mutex);
+            if (this->callbacks.size() == 0)
             {
                 throw std::bad_function_call();
             }
-            auto lastElement = callbacksSize - 1;
-            for (size_t i = 0; i < lastElement; i++)
+            for (auto callbackIt = this->callbacks.rbegin();
```

Comment:
Are you sure that iterator works faster for vector?

Reply:
I replaced std :: vector with std :: list, because we use the operations of adding an element to the end of the container and the operation of finding + removing an element from the container, as well as sequential traversal through the container. In the case of a vector, there was a problem with the asymptotics of removing an element in O (n). Since the vector was replaced with a list, we lost the ability to use operator []. Iterators, on the other hand, have almost no cost to use, so there is almost no difference between operator [] and using an iterator.

Reply:
Nice!